torch/csrc/jit/tensorexpr/ir_simplifier.h

```diff
@@ -9,17 +11,260 @@ namespace torch {
 namespace jit {
 namespace tensorexpr {
 
-// Uses the evaluator to fold an operation with constant terms.
-// Expr v must be evaluatable without Vars.
+// A bunch of helpers for determine the Dtype of the output of a multi argument
+// Term or Polynomial.
+namespace {
+template <class ExprType>
+Dtype promoteTypesVec(const Expr* s, std::vector<const ExprType*>& v) {
+  Dtype t = s->dtype();
+  bool first = true;
+
+  for (auto* e : v) {
+    if (first) {
+      t = Dtype(t.scalar_type(), e->dtype().lanes());
+      first = false;
+    }
+    t = promoteTypes(t, e->dtype());
+  }
+
+  return t;
+}
+
+template <class ExprType>
+Dtype promoteTypesVec(std::vector<const ExprType*>& v) {
+  if (v.empty()) {
+    throw malformed_input("empty list of types");
+  }
+
+  Dtype t = v[0]->dtype();
+  for (auto* e : v) {
+    t = promoteTypes(t, e->dtype());
+  }
+  return t;
+}
+
+template <class ExprType>
+Dtype promoteTypesMap(
+    const Expr* s,
+    std::unordered_map<SimplifierHashType, const ExprType*>& m) {
+  Dtype t = s->dtype();
+  bool first = true;
+  for (auto& e : m) {
+    if (first) {
+      t = Dtype(t.scalar_type(), e.second->dtype().lanes());
+      first = false;
+    }
+    t = promoteTypes(t, e.second->dtype());
+  }
+  return t;
+}
+
+template <class ExprType>
+Dtype promoteTypesVar(const ExprType* e) {
+  return e->dtype();
+}
+
+template <class ExprType, class... Args>
+Dtype promoteTypesVar(const ExprType* e, Args... es) {
+  Dtype lhs = e->dtype();
+  Dtype rhs = promoteTypesVar(es...);
+  if (e->isConstant()) {
+    lhs = Dtype(lhs.scalar_type(), rhs.lanes());
+  }
+
+  return promoteTypes(lhs, rhs);
+}
+
+// Helper for determining if an Expr is a multi-lane primitive (e.g. Broadcast
+// or Ramp).
+bool isMultilanePrimitive(const Expr* e) {
+  return e->expr_type() == IRNodeType::kBroadcast ||
+      e->expr_type() == IRNodeType::kRamp;
+}
+} // namespace
+
+// A Term represents a grouping of Exprs through multiplication.
+// E.g. product(scalar, *variables).
+class Term : public ExprNode<Term> {
+ public:
+  template <class... Args>
+  Term(HashProvider& hasher, const Expr* s, Args... ts)
+      : ExprNodeBase(promoteTypesVar(s, ts...)), scalar_(s), hasher_(hasher) {
+    CHECK(s->isConstant());
+    addComponent(ts...);
+    sort();
+  }
+
+  Term(HashProvider& hasher, const Expr* s, std::vector<const Expr*> v)
+      : ExprNodeBase(promoteTypesVec(s, v)),
+        variables_(std::move(v)),
+        scalar_(s),
+        hasher_(hasher) {
+    sort();
+  }
+
+  // Convenience constructor from a map of hash -> var, used when merging Terms.
+  Term(
+      HashProvider& hasher,
+      const Expr* s,
+      std::unordered_map<SimplifierHashType, const Expr*> varmap)
+      : ExprNodeBase(promoteTypesMap(s, varmap)), scalar_(s), hasher_(hasher) {
+    for (auto& p : varmap) {
```

Comment:
We're iterating over an unordered map here, which results in random order of components in the term. Would it be possible to fix the order somehow? Even if the order doesn't matter from computational standpoint, it would be annoying to debug this piece as the dumps could differ for no reason.

Reply:
The components are sorted below, so they end up in a consistent order.

Reply:
Or are you concerned about debugging the call to addComponent itself?

Reply:
> The components are sorted below, so they end up in a consistent order.
It's less of a problem then.

> Or are you concerned about debugging the call to addComponent itself?
I think it's fine to keep it as-is now then.